### PR TITLE
feat: Add UpdatePolicy and UpdatePolicies APIs

### PR DESCRIPTION
**Pull request by sociometry**
Fix: #182 

In addition, several new unit tests are also added, such as verifying return values, which Golang version seems not to cover. I've tried them with the Golang version and the results are basically the same.

### CODE REVIEW

**Comment by hsluoyz** [comment]
@sagilio

File: Casbin/Model/Assertion.cs [hunk 1]
```diff
@@ -46,28 +46,51 @@ public void RefreshPolicyStringSet()
             }
         }
 
-        internal void BuildIncrementalRoleLink(PolicyOperation policyOperation, IEnumerable<string> rule)
+        internal void BuildIncrementalRoleLink(PolicyOperation policyOperation, IEnumerable<string> rule, params IEnumerable<string>[] newRule)
```
Comment (sagilio) [inline Casbin/Model/Assertion.cs:49]: I think optional param is not clear enough, We can add a new overload like:

```csharp
internal void BuildIncrementalRoleLink(PolicyOperation policyOperation, IEnumerable<string> oldRule, IEnumerable<string>[] newRule)
```

File: Casbin/Model/Assertion.cs [hunk 2]
```diff
@@ -46,28 +46,51 @@ public void RefreshPolicyStringSet()
...
+                    BuildRoleLink(count, policyOperation, rule);
+                }
+            }
+            else
+            {
+                List<IEnumerable<string>> rulesList = rules.ToList();
+                List<IEnumerable<string>> newRulesList = newRule[0].ToList();
+                if (rulesList.Count != newRulesList.Count)
+                {
+                    return;
```
Comment (sagilio) [inline Casbin/Model/Assertion.cs:88]: Should throw an excption

File: Casbin/Model/Assertion.cs [hunk 3]
```diff
@@ -46,28 +46,51 @@ public void RefreshPolicyStringSet()
...
             {
-                BuildRoleLink(count, policyOperation, rule);
+                foreach (var rule in rules)
+                {
+                    BuildRoleLink(count, policyOperation, rule);
+                }
+            }
+            else
+            {
+                List<IEnumerable<string>> rulesList = rules.ToList();
```
Comment (sagilio) [inline Casbin/Model/Assertion.cs:84]:
```suggestion
                var rulesList = rules as IReadOnlyList<IEnumerable<string>> ?? rules.ToList();
```

File: Casbin.UnitTest/ModelTests/ManagementApiTest.cs [hunk 4]
```diff
@@ -122,6 +122,68 @@ public void TestModifyPolicy()
 
             e.RemoveFilteredPolicy(1, "");
             TestGetPolicy(e, AsList(AsList("eve", "data3", "read")));
+
+            bool res = e.UpdatePolicy(AsList("eve", "data3", "read"), AsList("eve", "data3", "write"));
```
Comment (sagilio) [inline Casbin.UnitTest/ModelTests/ManagementApiTest.cs:126]: We need to add similar tests into SyncModel dir.

**Review by sagilio — CHANGES_REQUESTED** [review]
Most changes looks good, but a bit issues need to change.

File: Casbin/Model/Assertion.cs [hunk 5]
```diff
@@ -86,7 +117,7 @@ public void BuildRoleLinks()
         }
 
         private void BuildRoleLink(int groupPolicyCount,
-            PolicyOperation policyOperation, IEnumerable<string> rule)
+            PolicyOperation policyOperation, IEnumerable<string> rule, params IEnumerable<string>[] newRule)
```
Comment (sagilio) [inline Casbin/Model/Assertion.cs:120]: remove `params

Comment (sagilio) [reply Casbin/Model/Assertion.cs:120]: I will fix later

**Review by sagilio — APPROVED** [review]
LGTM

**Pull request merged**